### ISSUE TITLE
Membership, storage and reminder providers as external extensions

### ISSUE THREAD

**Issue by ashkan-saeedi-mazdeh** [dotnet/orleans#2011]
I've realized the need to use a different storage provider for my application. 
I'm writing a complete storage provider, IMembershipTable and IReminderTable for CouchBase KVS. 
http://www.couchbase.com/

Should we in general consider moving these to Orleans like Azure, SQL and other cloud providers or keep them outside and just list them in contrib?

Also if in contributions, then should we include them in a visible part of documentation or not? We should really try the initiative of running Orleans everywhere for EVERYONE. As an unfortunate example in CouchBase's manual they mentioned Akka.NET in their async programming with .NET page but not Orleans and we should change this. I got offended :D 

What should be the approach @sergeybykov  @jdom  @ReubenBond  @galvesribeiro  Add others as you see fit.

**Comment by ashkan-saeedi-mazdeh** [comment]
Understood from @jdom  's comment on gitter when talking to @galvesribeiro  that the policy is to go to contributions.

**Comment by jdom** [comment]
Not necessarily. OrleansContrib can be an enabler to reach maturity, if it helps, but it is not the only option. Please read the thread at https://github.com/dotnet/orleans/issues/2005 for clarification on @galvesribeiro's conclusion.

I do believe that regardless of whether it's in the main repo or Orleans contrib, we should be doing a better effort to decouple each provider from the main solution. We should have some sort of test kit (as @veikkoeeva suggested) that we can run our providers against, to guarantee at least a minimal set of behavior from them. Of course each provider can then have additional provider specific tests.
Providers in the main repo should not have any advantage (in terms of how in-the-box they are, or how easy it is to test them) compared to new experimental/external providers.
Also it should be good if they can be versioned and released independently of Orleans core nugets.

**Comment by veikkoeeva** [comment]
@jdom The test kit idea could be realizable today with little work, at least for storage providers. We could create public data sets such as [example](https://github.com/dotnet/orleans/tree/master/test/TesterInternal/StorageTests/Relational/TestDataSets). Then create a core set of tests ([copy-paste-rename](https://github.com/dotnet/orleans/blob/master/test/TesterInternal/StorageTests/Relational/MySqlStorageTests.cs) ((this particular one inhertics for base ADO.NET one that isn't needed in general, I could fit ATS to that set too as for an example at some point if desired). The core tests intended to be stateless and test framework are [these](https://github.com/dotnet/orleans/blob/master/test/TesterInternal/StorageTests/Relational/CommonStorageTests.cs) that the aforementioned MySQL one sugars over with test sets and some ADO.NET/MySQL specific things.

This probably needs rearrangement in that regard. Also things like [this](https://github.com/dotnet/orleans/blob/master/test/TesterInternal/StorageTests/Relational/TestDataSets/GrainTypeGenerator.cs) could generate parameters using just a string template.

**Comment by galvesribeiro** [comment]
@veikkoeeva if you can make providers (stream, storage, reminder and membership at least) fully testable outside Orleans code, maybe issues like #2005 could be implemented correctly without the need to depend on the core. Let me know if you need any help as I don't want that things to happen again.

**Comment by ashkan-saeedi-mazdeh** [comment]
@galvesribeiro  exactly. as I pointed out in #2005 , the issues were mixed up and mostly this all happened because of lack of a technical capability (usability of tests for outside Orleans source), most of the times these issues happen due to management issues but this time was at least partially due to a technical one.

**Comment by shayhatsor** [comment]
The membership, reminders and statistics tests would also be easy moved into the test kit, it's basically copy pasting the test base classes (I do think that the liveness tests would need a bit of refactoring). How would you guys think this test kit would be packaged? as a nuget package, something like `Microsoft.Orleans.Providers.Tests` ?

**Comment by jdom** [comment]
That sounds great. My preference in naming would be `Microsoft.Orleans.Providers.TestKit`, but it's fine otherwise

**Comment by veikkoeeva** [comment]
@galvesribeiro
> if you can make providers (stream, storage, reminder and membership at least) fully testable outside Orleans code, maybe issues like #2005 could be implemented correctly without the need to depend on the core. Let me know if you need any help as I don't want that things to happen again. 

I added to my short note on how I envisoned it, the code supports it to some degree, I can point specifics in chat if someone asks or I'm around. I think the ADO.NET provider wen't largely to that direction (not fully, but closer) and I don't see why couldn't the other prociders too. It looks to me we should clean up the tests a bit. For instance, arrange more test arguments to test data generators and define a core set of tests that need to be satisfied and then concrete provider implementations could use these core sets and test data sets.

It's to some degree unrealised (looks like not much, I just go into that much specifics as it was inteded more as a spike) and perhaps a bit opinionated in that I would just like feed data in to functions and observe return values and side-effects. To arrange the "sandbox", the environment, could be realized too perhaps rather uniformly (how the ADO.NET provider gets the default arguments or gets them from outside and then sets up the environment, the factory methods perhaps could use inheritance, or then not).

My apologies for being sluggish here (happens in IRL and all that jazz).
